File: backend/app/schemas/backtests.py
from __future__ import annotations

from datetime import date
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class BacktestMarketRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    symbols: list[str] = Field(min_length=1, max_length=5)
    start_date: date = Field(alias="startDate")
    end_date: date = Field(alias="endDate")
    adjust: Literal["", "qfq", "hfq"] = "qfq"
    benchmark_symbol: str | None = Field(default=None, alias="benchmarkSymbol")

    @field_validator("symbols", mode="before")
    @classmethod
    def normalize_symbols(cls, value: Any) -> list[str]:
        if isinstance(value, str):
            items = value.split(",")
        else:
            items = value or []

        normalized = [str(item).strip() for item in items if str(item).strip()]
        if not normalized:
            raise ValueError("symbols must contain at least one stock code")
        return normalized


class BacktestStrategyRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    mode: Literal["preset", "custom"] = "custom"
    preset_id: str | None = Field(default=None, alias="presetId")
    code: str = ""
    params: dict[str, Any] = Field(default_factory=dict)

    @field_validator("code")
    @classmethod
    def normalize_code(cls, value: str) -> str:
        return value.strip()

    @model_validator(mode="after")
    def validate_mode(self) -> "BacktestStrategyRequest":
        if self.mode == "preset" and not self.preset_id:
            raise ValueError("preset 模式下必须提供 presetId")
        if self.mode == "custom" and not self.code:
            raise ValueError("custom 模式下必须提供 strategyCode")
        return self


class BacktestExecutionRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    mode: Literal["close", "next_open"] = "close"
    position_size: float = Field(default=1.0, alias="positionSize", gt=0, le=1)
    lot_size: int = Field(default=100, alias="lotSize", ge=1)


class BacktestCostRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    trading_fee: float = Field(default=0.0005, alias="tradingFee", ge=0)
    stamp_tax: float = Field(default=0.001, alias="stampTax", ge=0)
    slippage: float = Field(default=0.0, ge=0)


class BacktestRiskRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    stop_loss: float = Field(default=0.0, alias="stopLoss", ge=0)
    take_profit: float = Field(default=0.0, alias="takeProfit", ge=0)


class BacktestRunRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    market: BacktestMarketRequest
    strategy: BacktestStrategyRequest
    execution: BacktestExecutionRequest = Field(default_factory=BacktestExecutionRequest)
    costs: BacktestCostRequest = Field(default_factory=BacktestCostRequest)
    risk: BacktestRiskRequest = Field(default_factory=BacktestRiskRequest)
    initial_capital: float = Field(default=100000.0, alias="initialCapital", gt=0)

    @model_validator(mode="before")
    @classmethod
    def support_legacy_payload(cls, payload: Any) -> Any:
        if not isinstance(payload, dict):
            return payload
        if "market" in payload or "strategy" in payload:
            return payload

        return {
            "market": {
                "symbols": payload.get("symbols"),
                "startDate": payload.get("startDate"),
                "endDate": payload.get("endDate"),
                "adjust": payload.get("adjust", "qfq"),
                "benchmarkSymbol": payload.get("benchmarkSymbol"),
            },
            "strategy": {
                "mode": payload.get("strategyMode", "custom"),
                "presetId": payload.get("strategyPreset"),
                "code": payload.get("strategyCode", ""),
                "params": payload.get("params", {}),
            },
            "execution": {
                "mode": payload.get("executionMode", "close"),
                "positionSize": payload.get("positionSize", 1.0),
                "lotSize": payload.get("lotSize", 100),
            },
            "costs": {
                "tradingFee": payload.get("tradingFee", 0.0005),
                "stampTax": payload.get("stampTax", 0.001),
                "slippage": payload.get("slippage", 0.0),
            },
            "risk": {
                "stopLoss": payload.get("stopLoss", 0.0),
                "takeProfit": payload.get("takeProfit", 0.0),
            },
            "initialCapital": payload.get("initialCapital", 100000.0),
        }

    @model_validator(mode="after")
    def validate_dates(self) -> "BacktestRunRequest":
        if self.market.end_date < self.market.start_date:
            raise ValueError("endDate must be on or after startDate")
        return self

    @property
    def symbols(self) -> list[str]:
        return self.market.symbols

    @property
    def start_date(self) -> date:
        return self.market.start_date

    @property
    def end_date(self) -> date:
        return self.market.end_date

    @property
    def adjust(self) -> str:
        return self.market.adjust

    @property
    def benchmark_symbol(self) -> str | None:
        return self.market.benchmark_symbol

    @property
    def strategy_code(self) -> str:
        return self.strategy.code

    @property
    def strategy_mode(self) -> str:
        return self.strategy.mode

    @property
    def strategy_preset(self) -> str | None:
        return self.strategy.preset_id

    @property
    def params(self) -> dict[str, Any]:
        return self.strategy.params

    @property
    def execution_mode(self) -> str:
        return self.execution.mode

    @property
    def position_size(self) -> float:
        return self.execution.position_size

    @property
    def lot_size(self) -> int:
        return self.execution.lot_size

    @property
    def trading_fee(self) -> float:
        return self.costs.trading_fee

    @property
    def stamp_tax(self) -> float:
        return self.costs.stamp_tax

    @property
    def slippage(self) -> float:
        return self.costs.slippage

    @property
    def stop_loss(self) -> float:
        return self.risk.stop_loss

    @property
    def take_profit(self) -> float:
        return self.risk.take_profit

File: frontend/lib/backtests.ts
import type { BacktestRunPayload, ExecutionMode, StrategyMode } from "@/types/api";
import { displayExecutionMode, displayRatioPercent } from "@/lib/display";

export type BacktestFormValues = {
  symbolsInput: string;
  startDate: string;
  endDate: string;
  benchmarkSymbol: string;
  strategyMode: StrategyMode;
  strategyPreset: string;
  strategyCode: string;
  initialCapital: number;
  executionMode: ExecutionMode;
  positionSize: number;
  lotSize: number;
  volumeLimitPct: number;
  tradingFee: number;
  stampTax: number;
  slippage: number;
  minCommission: number;
  transferFeeRate: number;
  adjust: "" | "qfq" | "hfq";
  stopLoss: number;
  takeProfit: number;
  maxDrawdown: number;
  maxDailyLoss: number;
  maxPositionSize: number;
  reduceOnlyAfterRisk: boolean;
  riskCooldownBars: number;
  params: Record<string, unknown>;
};

export type BacktestQuickProfileId =
  | "research_default"
  | "cost_aware"
  | "trend_following";

export type BacktestQuickProfile = {
  id: BacktestQuickProfileId;
  label: string;
  description: string;
  values: Pick<
    BacktestFormValues,
    | "executionMode"
    | "positionSize"
    | "lotSize"
    | "volumeLimitPct"
    | "tradingFee"
    | "stampTax"
    | "slippage"
    | "minCommission"
    | "transferFeeRate"
    | "stopLoss"
    | "takeProfit"
    | "maxDrawdown"
    | "maxDailyLoss"
    | "maxPositionSize"
    | "reduceOnlyAfterRisk"
    | "riskCooldownBars"
  >;
};

export type BacktestSummaryItem = {
  label: string;
  value: string;
};

export const BACKTEST_QUICK_PROFILES: BacktestQuickProfile[] = [
  {
    id: "research_default",
    label: "研究默认",
    description: "接近当前研究台默认设置，适合先快速看策略轮廓。",
    values: {
      executionMode: "close",
      positionSize: 1,
      lotSize: 100,
      volumeLimitPct: 0.25,
      tradingFee: 0.0005,
      stampTax: 0.001,
      slippage: 0,
      minCommission: 5,
      transferFeeRate: 0.00002,
      stopLoss: 0,
      takeProfit: 0,
      maxDrawdown: 0,
      maxDailyLoss: 0,
      maxPositionSize: 0,
      reduceOnlyAfterRisk: false,
      riskCooldownBars: 0,
    },
  },
  {
    id: "cost_aware",
    label: "保守成交",
    description: "更偏重成交摩擦和风险约束，适合先做现实性校验。",
    values: {
      executionMode: "next_open",
      positionSize: 0.75,
      lotSize: 100,
      volumeLimitPct: 0.15,
      tradingFee: 0.0003,
      stampTax: 0.001,
      slippage: 0.0002,
      minCommission: 5,
      transferFeeRate: 0.00002,
      stopLoss: 0.05,
      takeProfit: 0.16,
      maxDrawdown: 0.12,
      maxDailyLoss: 800,
      maxPositionSize: 0,
      reduceOnlyAfterRisk: true,
      riskCooldownBars: 2,
    },
  },
  {
    id: "trend_following",
    label: "趋势放大",
    description: "偏向中强趋势研究，持仓更积极，容忍更大波动。",
    values: {
      executionMode: "next_open",
      positionSize: 0.95,
      lotSize: 100,
      volumeLimitPct: 0.25,
      tradingFee: 0.0005,
      stampTax: 0.001,
      slippage: 0.0005,
      minCommission: 5,
      transferFeeRate: 0.00002,
      stopLoss: 0.08,
      takeProfit: 0.24,
      maxDrawdown: 0.18,
      maxDailyLoss: 0,
      maxPositionSize: 0,
      reduceOnlyAfterRisk: false,
      riskCooldownBars: 0,
    },
  },
];

export function parseSymbolsInput(value: string): string[] {
  return value
    .split(",")
    .map((item) => item.trim())
    .filter(Boolean)
    .slice(0, 5);
}

export function validateBacktestForm(values: BacktestFormValues): string[] {
  const errors: string[] = [];
  if (!parseSymbolsInput(values.symbolsInput).length) {
    errors.push("请至少输入 1 个股票代码。");
  }

  if (values.startDate && values.endDate && values.endDate < values.startDate) {
    errors.push("结束日期不能早于开始日期。");
  }

  if (values.strategyMode === "preset" && !values.strategyPreset) {
    errors.push("预设模式下必须选择一个策略预设。");
  }

  if (values.strategyMode === "custom" && !values.strategyCode.trim()) {
    errors.push("自定义模式下必须提供策略代码。");
  }

  return errors;
}

export function applyQuickProfile(
  values: BacktestFormValues,
  profileId: BacktestQuickProfileId,
): BacktestFormValues {
  const profile = BACKTEST_QUICK_PROFILES.find((item) => item.id === profileId);
  if (!profile) {
    return values;
  }

  return {
    ...values,
    ...profile.values,
  };
}

export function buildBacktestPayload(values: BacktestFormValues): BacktestRunPayload {
  return {
    market: {
      symbols: parseSymbolsInput(values.symbolsInput),
      startDate: values.startDate,
      endDate: values.endDate,
      adjust: values.adjust,
      benchmarkSymbol: values.benchmarkSymbol || undefined,
    },
    strategy: {
      mode: values.strategyMode,
      presetId: values.strategyPreset || undefined,
      code: values.strategyCode,
      params: values.params,
    },
    execution: {
      mode: values.executionMode,
      positionSize: values.positionSize,
      lotSize: values.lotSize,
      volumeLimitPct: values.volumeLimitPct,
    },
    costs: {
      tradingFee: values.tradingFee,
      stampTax: values.stampTax,
      slippage: values.slippage,
      minCommission: values.minCommission,
      transferFeeRate: values.transferFeeRate,
    },
    risk: {
      stopLoss: values.stopLoss,
      takeProfit: values.takeProfit,
      maxDrawdown: values.maxDrawdown,
      maxDailyLoss: values.maxDailyLoss,
      maxPositionSize: values.maxPositionSize,
      reduceOnlyAfterRisk: values.reduceOnlyAfterRisk,
      riskCooldownBars: values.riskCooldownBars,
    },
    initialCapital: values.initialCapital,
  };
}

export function buildBacktestFormSummary(
  values: BacktestFormValues,
  strategyLabel?: string,
): BacktestSummaryItem[] {
  const symbols = parseSymbolsInput(values.symbolsInput);
  return [
    {
      label: "标的范围",
      value: `${symbols.length} 个标的 / 基准 ${values.benchmarkSymbol || "未设置"}`,
    },
    {
      label: "策略",
      value: `${strategyLabel || values.strategyPreset || "自定义策略"} / ${
        values.strategyMode === "preset" ? "预设模式" : "自定义模式"
      }`,
    },
    {
      label: "执行",
      value: `${displayExecutionMode(values.executionMode)} / 仓位 ${(
        values.positionSize * 100
      ).toFixed(0)}% / 成交量限制 ${toPercent(values.volumeLimitPct)}`,
    },
    {
      label: "成本与风控",
      value: `佣金 ${toPercent(values.tradingFee)} / 滑点 ${toPercent(values.slippage)} / 止损 ${
        values.stopLoss > 0 ? toPercent(values.stopLoss) : "关闭"
      } / 回撤阈值 ${values.maxDrawdown > 0 ? toPercent(values.maxDrawdown) : "关闭"}`,
    },
  ];
}

export function formatMetricValue(value: number | null | undefined): string {
  return displayRatioPercent(value);
}

export function formatDecimal(value: number | null | undefined, digits = 2): string {
  if (typeof value !== "number" || Number.isNaN(value)) {
    return "暂无";
  }
  return value.toFixed(digits);
}

function toPercent(value: number): string {
  return `${(value * 100).toFixed(2)}%`;
}
